Feature/issue#13 플레이리스트 등록, 수정, 조회, 삭제 기능 구현

kdmstj opened the pull request:
### ✍️ Description
- 플레이리스트 등록, 수정, 조회(플레이리스트 상세 조회, 목록 조회), 삭제 기능을 구현하였습니다. 

<br>

### 🎯 Key Changes
- DTO <-> Entity 를 매핑하는 MapStruct 라이브러리를 사용하였습니다. 
   - 기존에는 엔티티 그대로 Controller에게 반환하였지만, 이에 대해서 필요하지 않은 정보까지 제공한다는 점에서 위험 요소가 된다고 판단하여 Service 계층에서 Mapper에게 entity를 dto로 변환한 것을 위임하였습니다. 
   - MapStruct가 생성한 코드는 다음 디렉토리에 있습니다. `build/generated/sources/annotationProcessor/java/main/com/petitapetit/miml/domain/playlist/mapper/`
- `UriComponentsBuilder`를 통해서 자원이 생성되었을 때 URI를 생성해주는 유틸을 만들어 역할을 분리하였습니다.
- `GlobalExceptionHandler` 에서 @RestControllerAdvice 어노테이션을 통해 여러 컨트롤러에 대해 발생하는 예외에 ExceptionHandler가 적용되도록 하였습니다. 

### 💬 To Reviewers
- `PlayListService` 의 `checkAuthorization`과 `checkExistence`에서 발생하는 예외에 대해서 어떤 상태 코드를 줘야 할지에 대해 의견을 여쭤보고 싶습니다. (해당 코드에 작성된 상태코드는 임의로 작성해놓았습니다.)

kdmstj committed:
test: 서비스 플레이리스트 저장 로직 테스트 코드 작성

- 서비스단에서 플레이리스트 등록 로직을 확인하기 위하여 코드를 작성하였습니다.
- 요청에 따라 플레이리스트를 저장할 엔티티를 생성해주었습니다.(PlayList)
- 플레이리스트를 등록하는 레포지토리를 생성해주었습니다.(PlayListRepository)
- 서비스 계층에서 플레이리스트를 등록하는 메소드를 생성해주었습니다.(savePlayList)

kdmstj committed:
Merge branch 'develop' of https://github.com/petit-a-petit/MIML into feature/3

kdmstj committed:
fix: 플레이리스트 공개 여부 변수명 변경

- 플레이리스트 공개 여부 변수명에 대해 boolean 변수이므로 be동사 is를 사용하여 ture/false를 표현하는것이 좋다고 생각되어 변경되었습니다.

kdmstj committed:
fix: 플레이리스트 등록 요청 DTO 변수명 변경

- 기존에 플레이리스트 등록 요청 DTO 변수명에 대해 HTTP 메소드가 포함되지 않는 것이 권장됨을 알게되어 변경하게 되었습니다.

kdmstj committed:
build: mapstruct 라이브러리 추가

- 서로 다른 객체 간 매핑을 자동으로 해주는 mapstruct 라이브러리 추가하였습니다.
- lombok 라이브러리를 사용하기 때문에 그 후에 추가하였습니다.

kdmstj committed:
refactor: mapstruct 라이브러리를 통한 dto<->entity 매핑

- PlayListService 에서 각 모델간의 매핑을 해주는 역할을 가진 PlayListMapper 의존성 주입을 받았습니다.
- playListMapper 에 요청받은 dto를 entity로 변환할 것을 위임하였습니다.

kdmstj committed:
refactor: PlayList 등록에 대한 응답 DTO 생성

- PlayList 등록에 대한 응답값으로 기존에는 엔티티 그대로 controller 에게 반환하였지만, 이에 대해서 필요하지 않는 정보까지 제공한다는 점에서 위험 요소가 됩니다.
- 이를 해결하고자 등록된 playListId 를 담는 PlayListDto.SaveResponse를 생성하였습니다.
- PlayListService에서 PlayListMapper에게 dto를 entity로 변환할 것을 위임하였습니다.

kdmstj committed:
test: PlayList 등록 테스트 코드

- 플레이리스트 등록 단위 테스트(Controller, Service Layer)를 작성하였습니다.

kdmstj committed:
fix: build.gradle 충돌 해결

- mapstruct 의존성 설정과 이전 설정과의 충돌을 해결하였습니다.

kdmstj committed:
feat: Uri 생성 유틸 생성

- 자원이 생성되었을 때 URI를 생성해주는 유틸을 만들어 역할을 분리하였습니다.
- UriComponentsBuilder를 통해서 구현하였습니다.

kdmstj committed:
feat: Controller 플레이리스트 수정/조회/삭제 엔드포인트 구현

- 특정 플레이리스트에 접근해야 하는 경우 (수정, 특정 플레이리스트 조회, 삭제) @PathVariable 을 사용하였습니다.
- ResponseEntity 로 HttpStatus 코드와 Body 를 반환하였습니다.

kdmstj committed:
feat: PlayList 요청/응답 DTO 생성

- 등록/수정에 대한 요청/응답은 SaveRequest/SaveResponse 정적 클래스를 사용하였습니다.
- 플레이리스트 목록/상세 조회에 대한 응답은 Response 정적 클래스를 사용하였습니다. 추후에 등록된 노래 목록들과 플레이리스트 등록한 사람의 이름을 함께 응답할 예정입니다.

kdmstj committed:
feat: Service 플레이리스트 수정/조회/삭제 비즈니스 로직 구현

- MapStruct 라이브러리를 사용하여 DTO<->Entity 매핑을 하였습니다. MapStruct로 인하여 생성된 코드는 다음과 같습니다.
build/generated/sources/annotationProcessor/java/main/com/petitapetit/miml/domain/playlist/mapper/PlayListMapperImpl.java
- 존재하는 플레이리스트인지, 플레이리스트에 대한 권한이 있는지에 대해 (getVerifiedPlayList())메소드로 트랜잭션 처리를 해주었습니다.
- 플레이리스트 수정 시 존재하는 플레이리스트에 대해 필드값을 변경하도록 Setter 어노테이션을 추가하였습니다.
- 추후에 예외처리를 정리할 예정입니다.

kdmstj committed:
feat: 예외 공통 처리 기능 추가

- GlobalExceptionHandler 클래스에서 @RestControllerAdvice 어노테이션을 통해 여러 @RestController 어노테이션이 붙은 컨트롤러에 대해 ExceptionHandler가 적용되도록 하였습니다.
- 스프링 예외를 미리 정의한 ResponseEntityExceptionHandler 추상 클래스를 상속받아 예외 처리기를 동일하게 하여 (스프링 예외들은 DefaultHandelrExcpetionResolver 가 처리함.) 클라이언트가 일관된 에러 응답을받도록 하였습니다.
- handlerExceptionInternal 을 오버라이딩 하여 스프링 예외 대한 에러 응답을 보내주었습니다.
- ErrorCode 인터페이스를 구현하는 CommonErrorCode enum을 생성하였습니다. (추후에 도메인 별 ErrorCode 생성을 위해 네이밍 시 Common을 붙였습니다.)
- 클라이언트에게 공통으로 보내줄 ErrorResponse 클래스를 생성하였습니다.
- PlayListService 에서 발생하는 언체크 예외에 대한 상태 코드는 추후 회의를 통해 변경할 예정입니다.

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/playlist/controller/PlayListController.java:
```diff
@@ -1,19 +1,67 @@
+        return ResponseEntity.created(uri).body(response);
+    }
+
+    @PatchMapping("/{play_list_id}")
+    public ResponseEntity<PlayListDto.SaveResponse> patchPlayList(@PathVariable("play_list_id") Long playListId,
+                                                                  @RequestBody PlayListDto.SaveRequest request){
+
+        Long memberId = 1L;//mock
+        PlayListDto.SaveResponse response = playListService.modifyPlayList(playListId, request, memberId);
+
+        return ResponseEntity.ok(response);
+    }
```
`NIT` : 요건 단순한 팁인데 단순히 `JSON` + `HttpStatus.OK` 라면 ResponseEntity 를 Wrapping 하지 않으셔도 됩니다. 일관성이 있어 보기 좋네요 😀
```suggestion
    @PatchMapping("/{play_list_id}")
    public PlayListDto.SaveResponse patchPlayList(@PathVariable("play_list_id") Long playListId,
                                                                  @RequestBody PlayListDto.SaveRequest request){

        Long memberId = 1L;//mock
        return playListService.modifyPlayList(playListId, request, memberId);
    }
```

kdmstj replied:
넵 참고하도록 하겠습니다 :)

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/playlist/controller/PlayListController.java:
```diff
@@ -1,19 +1,67 @@
+
+        URI uri = UriUtil.createUri(DEFAULT_URI, response.getPlayListId());
+
+        return ResponseEntity.created(uri).body(response);
```
오.. 경로를 이렇게도 담아줄 수 있나보군요. 배워갑니다👍

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/playlist/service/PlayListService.java:
```diff
@@ -1,7 +1,89 @@
+    }
+
+    private void checkAuthorization(Long memberId, PlayList playList) {
+        if(!Objects.equals(playList.getPlayListId(), memberId)){
```
`playList.getPlayListId()`, `memberId` 와 동등성 비교를 하네요?? 혹시 이유가 있나요?🧐 PlayList에 따로 memberId가 있는것으로 확인했거든요!

kdmstj replied:
memberId를 비교해야 하는 데 잘못 작성한 부분이 있었네요!!
캐치해주셔서 감사합니다 :)

Nine-JH commented on src/main/java/com/petitapetit/miml/global/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,13 @@
+
+@Builder @Getter
+@RequiredArgsConstructor
+public class ErrorResponse {
+
+    private final String code;
+    private final String message;
+}
```
해당문제가 언제 발생했는지에 대한 `TimeStamp`도 찍었으면 좋을 것 같아요! 프론트엔드와 협업 당시에 로그 찾는것도 일이었던 경험이 있어서 ㅋㅋㅋ

kdmstj replied:
해당부분에 대해 너무 공감합니다!!ㅎㅎ 추가하도록 하겠습니다~!

Nine-JH commented on src/main/java/com/petitapetit/miml/global/exception/GlobalExceptionHandler.java:
Application Exception을 분리한게 인상적이네요 👍👍👍👍

Nine-JH commented on src/main/java/com/petitapetit/miml/util/UriUtil.java:
```diff
@@ -0,0 +1,14 @@
+import java.net.URI;
+
+public class UriUtil {
+    public static URI createUri(String defaultUrl, Long resourceId) {
```
`NIT`: `resourceId`도 좋아보이는데 조금 더 넓은 의미로 `uriVariables` `pathVariable`도 괜찮아 보이네요

kdmstj replied:
넵 추후에 사용할 것을 고려하여 uriVariable로 수정하도록 하겠습니다 !!

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/playlist/service/PlayListService.java:
```diff
@@ -1,7 +1,89 @@
+            playList.setName(request.getName());
+        }
+
+        PlayList modifiedPlayList = playListRepository.save(playList);
```
수정의 경우 JPA 에서는 더티 체킹을 수행하기 때문에 따로 `save()` 메서드를 사용하지 않아도 저장이 됩니다 😀 

https://jojoldu.tistory.com/415

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/playlist/service/PlayListService.java:
```diff
@@ -1,7 +1,89 @@
+        playListRepository.delete(playList);
+    }
+
+    @Transactional
+    public PlayList getVerifiedPlaylist(Long playListId, Long memberId) {
+
+        PlayList playList = checkExistence(playListId);
+        checkAuthorization(memberId, playList);
+
+        return playList;
+    }
```
클래스 레벨에서 트랜잭션 관리를 해주는 것도 좋아보이네요.
그리고 `getVerifiedPlaylist` 만 `@Transactional` 설정을 하신 이유가 있을까요?

kdmstj replied:
넵 말씀하신대로 클래스 레벨에서 트랜잭션 관리를 해주는 것이 좋다고 판단하여 코드 변경하였습니다!

Nine-JH reviewed (commented):
넘모넘모 고생하셨습니다 👍👍👍
여쭤보신거에 대한 저의 생각은 요렇게 입니다.
* checkAuthorization -> 403 Error
* checkExistence -> 404 Error (Not Found는 일관된 데이터를 주기 때문에 캐시가 가능함)

minju412 commented on src/main/java/com/petitapetit/miml/domain/playlist/controller/PlayListController.java:
```diff
@@ -1,19 +1,67 @@
+        return ResponseEntity.created(uri).body(response);
+    }
+
+    @PatchMapping("/{play_list_id}")
```
수정 메서드에 대해 PUT 방식 대신 PATCH 방식을 선택하신 기준이 궁금합니다!

kdmstj replied:
1. PUT의 메서드의 규약을 따르기 위해서는 존재하지 않는 플레이리스트에 대해 수정을 요청한 경우 새로운 플레이리스트를 추가해주어야 합니다. 하지만, 개발 요구사항에 의해 존재하는 플레이리스트를 수정해야 하기 때문에 PATCH가 더 적절하였습니다.
2. 수정하고 싶은 데이터(플레이리스트 이름, 공개 여부)에 대해 부분적으로 수정하는 요구사항에 따라 대상 리소스를 나타내는 데이터를 대체하는 PUT 메서드보다 부분적인 수정을 하는 PATCH 메서드가 더 적합하다고 판단하였습니다.

minju412 replied:
혹시 1번의 경우 존재하지 않는 플레이리스트에 대해 수정을 요청한다면 예외 상황으로 처리하는 것에 대해서는 어떻게 생각하시나요?!

minju412 commented on src/main/java/com/petitapetit/miml/domain/playlist/dto/PlayListDto.java:
```diff
@@ -1,14 +1,34 @@
+
+    @AllArgsConstructor @NoArgsConstructor
+    @Builder @Getter
+    public static class Response {
```
`SaveResponse` 처럼 해당 Response도 어떨 때 사용되는 Response인지 이름을 통해 나타내면 좋을 것 같습니다!

minju412 commented on src/main/java/com/petitapetit/miml/domain/playlist/entity/PlayList.java:
```diff
@@ -0,0 +1,25 @@
+
+    private String name;
+
+    private Long memberId;
```
Member와 PlayList의 관계가 N:1 일것으로 생각되는데, 연관관계를 설정하지 않으신 이유가 있을까요?!

minju412 replied:
앗 지금 생각해보니 mock 데이터를 담기 위해서일까요..??

kdmstj replied:
넵 맞습니다! 빠른 개발을 위해서 User와 아직 연관관계를 설정해두지 않았습니다!

minju412 commented on src/main/java/com/petitapetit/miml/domain/playlist/service/PlayListService.java:
```diff
@@ -1,7 +1,89 @@
+        PlayList playList = getVerifiedPlaylist(playListId, memberId);
+
+        if (request.getIsPublic() != null) {
+            playList.setIsPublic(request.getIsPublic());
```
setter 대신 캡슐화를 적용하여 메시지(`playList.updateIsPublic(request.getIsPublic());`)를 통해 객체의 상태를 변경하면 더욱 객체지향적인 설계가 될 것 같아요!

```java
public class PlayList {
    ...
    public void updateIsPublic(Boolean isPublic) {
    this.isPublic = isPublic;
    }
}
```

kdmstj replied:
넵!! 저도 동의합니다ㅎㅎ

minju412 commented on src/main/java/com/petitapetit/miml/domain/playlist/service/PlayListService.java:
```diff
@@ -1,7 +1,89 @@
+        }
+
+        if(request.getName() != null){
+            playList.setName(request.getName());
```
위와 마찬가지입니다!

minju412 commented on src/main/java/com/petitapetit/miml/domain/playlist/service/PlayListService.java:
```diff
@@ -1,7 +1,89 @@
+    }
+
+    private PlayList checkExistence(Long PlayListId) {
+        Optional<PlayList> optional = playListRepository.findById(PlayListId);
```
중요한건 아니지만, 86, 87번 라인을 합쳐 아래처럼 사용할 수 있을 것 같습니다! 
`return playListRepository.findById(PlayListId).orElseThrow(()-> new RestApiException(CommonErrorCode.RESOURCE_NOT_FOUND));`

minju412 commented on src/main/java/com/petitapetit/miml/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface ErrorCode {
+
+    String name();
```
`ErrorCode`에 `name`에는 예를 들어 어떤 값이 담길까요..?!

kdmstj replied:
ErrorCode의 name에는 에러 이름을 담으며, (ex. RESOURCE_NOT_FOUND)
ErrorResponse 의 body 생성 시 ErrorCode.name() 메소드를 통해 에러 이름을 담는데 사용됩니다.
```
CommonErrorCode.java

RESOURCE_NOT_FOUND(HttpStatus.NOT_FOUND, "Resource not exists");
```
ErrorCode 인터페이스를 구현한 CommonErrorCode 열거형에서는 name() 메소드로 열거 객체의 문자열을 리턴합니다.
열거 상수는 에러 이름으로 선언하였기 때문에 객체로 생성 후 문자열을 리턴하게 되면, 에러 이름이 반환됩니다.

minju412 commented on src/main/java/com/petitapetit/miml/util/UriUtil.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.net.URI;
+
+public class UriUtil {
```
우재님께서 알려주신건데, lombok의 `@UtilityClass` 어노테이션에 대해 알아보셔도 좋을 것 같습니다!

minju412 reviewed (changes requested):
정말 고생하셨습니다! 👍
여쭤봐 주신 것에 대해서는 저도 자현님의 의견에 동의합니다!

Jeonwoojae commented on src/main/java/com/petitapetit/miml/global/exception/RestApiException.java:
```diff
@@ -0,0 +1,11 @@
+
+@RequiredArgsConstructor
+@Getter
+public class RestApiException extends RuntimeException{
```
예외 사항을 커스텀 하는건 매우 좋습니다!👍
하지만 현재 사용하고 있는 부분이 RestApiException 보다는 내부 로직에서 발생하는 예외라 다른 이름이면 이해하기 편할 것 같습니다!
팀에서 회의를 통해 예외를 어떻게 처리할지 전략을 정하면 좋을 것 같네요😁

kdmstj replied:
넵 좋습니다!!ㅎㅎ RestApiException 네이밍에 대해서는 더 고민해보겠습니다🤗 -> BusinessException

Jeonwoojae reviewed (commented):
고생하셨습니다👍
Mapper를 잘 몰라서 보고 배울 수 있도록 하겠습니다✍

kdmstj committed:
refactor: 공통 예외 처리

- 내부 로직을 처리하기 때문에 BusinessException 으로 네이밍을 변경하였습니다.
- 플레이리스트에 대한 권한이 없는 경우 403,플레이리스트가 없는 경우 404 HTTP 상태코드를 응답으로 주도록 하였습니다.
- 에러 응답에 대해 에러가 언제 발생했는지에 대해 나타낼 수 있는 TimeStamp 를 추가하였습니다.

kdmstj committed:
refactor: PlayList 수정 메서드 추가

- 역할을 분리하기 위해서 PlayList 에 대한 수정 사항(name, isPublic)을 PlayList 에서 하도록 메서드를 생성하였습니다.

kdmstj committed:
refactor: mapper 메서드 이름 변경

- 자바 메서드 명명 규칙에 따라 메서드 이름 첫번째 단어를 소문자로 작성하였습니다.

kdmstj committed:
refactor: 플레이리스트 응답 DTO 네이밍 변경

- 플레이리스트 상세 정보에 대한 응답임을 명시하고자 DetailResponse 로 클래스명을 변경하였습니다.

kdmstj committed:
refactor: UriUtil 유틸리티 어노테이션 추가

- 유틸리티 어노테이션 추가를 통해 해당 클래스가 유틸리티 클래스임을 더 알아보기 쉽도록 하였습니다.
- 추후 확장성을 위해 resourceId 에서 uriPathVariable 로 변경하였습니다.

kdmstj committed:
refactor: PlayListService 코드 변

- 클래스 레벨로 트랜잭션 관리를 하도록 하였습니다.
- 플레이리스트 수정시 Optional 로 요청사항에 대한 null 값을 확인하도록 하였습니다.
- checkAuthorization 메소드에서 플레이리스트 소유자인지 확인하는 코드로 변경하였습니다.

kdmstj committed:
refactor: 플레이리스트 상세 정보 응답 클래스 이름 변경

- 플레이리스트에 대한 상세 정보 응답 클래스에 대해 더 알아보기 쉽도록 이름을 변경하였습니다.

minju412 reviewed (commented):
코멘트가 꼼꼼하게 반영된 것 같습니다!! 고생하셨습니다!! 😊
